### DISCUSSION TITLE
Adding Folia Support

### DISCUSSION THREAD

**Discussion by GeorgeV220** [denmeh/holoeasy#70] [Ideas]
The main downside is that we need Java 21, which is required for both the Folia API and the newer iterations of the Paper API. 

To support older versions, we could consider leveraging [JvmDowngrader](https://github.com/unimined/JvmDowngrader) as a potential solution for maintaining compatibility with earlier Java versions. However, it is important to note that this route could introduce additional complexities, particularly regarding the publishing of the downgraded JAR on JitPack.

Alternatively, I suggest structuring the project so that the core module remains compatible with Java 1.8 while implementing schedulers as separate modules. The Bukkit scheduler would serve as the default, with the Folia scheduler available as an optional addition.

**Comment by GeorgeV220** [comment]
Another solution is to drop 1.8-1.16.5, there are forks of paper/spigot that work with Java 17/21 but plugins won't play nicely with newer java versions

**Comment by denmeh** [comment]
We could create an interface on the holoeasy-core module and create other module as spigot and folia with their custom implementation.
But we have to think also if there is demand for Folia, because we have to maintain that api.

**Reply by GeorgeV220** [reply]
The API will work as-is with Folia; the only thing holding us back now is the scheduler. I have some temporary changes in place, but I was considering adding an interface to the core along with a Bukkit scheduler implementation, then creating a Folia scheduler module (users would need to add it as a dependency to work with Folia, though).

**Reply by GeorgeV220** [reply]
Examples:
https://github.com/GeorgeV220/MartexLibrary/tree/main/minecraft/shared/src/main/java/com/georgev22/library/minecraft/scheduler

https://github.com/GeorgeV220/MartexLibrary/tree/main/minecraft/bukkit/src/main/java/com/georgev22/library/minecraft/scheduler

the shared scheduler and bukkit can be in the core and the folia one on a different module because it requires java 17/21
